# Retail Sales Prediction

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Problem-Statement" data-toc-modified-id="Problem-Statement-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Problem Statement</a></span><ul class="toc-item"><li><span><a href="#Objective" data-toc-modified-id="Objective-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Objective</a></span></li></ul></li><li><span><a href="#Data-description" data-toc-modified-id="Data-description-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Data description</a></span><ul class="toc-item"><li><span><a href="#Stores" data-toc-modified-id="Stores-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Stores</a></span></li><li><span><a href="#Features" data-toc-modified-id="Features-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Features</a></span></li><li><span><a href="#Sales" data-toc-modified-id="Sales-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Sales</a></span></li></ul></li><li><span><a href="#Data-Exploration" data-toc-modified-id="Data-Exploration-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Data Exploration</a></span><ul class="toc-item"><li><span><a href="#Importing-and-Reading-the-datasets" data-toc-modified-id="Importing-and-Reading-the-datasets-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Importing and Reading the datasets</a></span><ul class="toc-item"><li><span><a href="#Initial-thoughts" data-toc-modified-id="Initial-thoughts-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Initial thoughts</a></span><ul class="toc-item"><li><span><a href="#Steps-to-merge-the-datasets" data-toc-modified-id="Steps-to-merge-the-datasets-3.1.1.1"><span class="toc-item-num">3.1.1.1&nbsp;&nbsp;</span>Steps to merge the datasets</a></span></li><li><span><a href="#Final-dataset" data-toc-modified-id="Final-dataset-3.1.1.2"><span class="toc-item-num">3.1.1.2&nbsp;&nbsp;</span>Final dataset</a></span></li></ul></li></ul></li><li><span><a href="#Save-to-file" data-toc-modified-id="Save-to-file-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Save to file</a></span><ul class="toc-item"><li><span><a href="#CSV" data-toc-modified-id="CSV-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>CSV</a></span></li><li><span><a href="#Pickle" data-toc-modified-id="Pickle-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Pickle</a></span></li></ul></li><li><span><a href="#Data-description" data-toc-modified-id="Data-description-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Data description</a></span></li><li><span><a href="#Univariate-Analysis" data-toc-modified-id="Univariate-Analysis-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Univariate Analysis</a></span><ul class="toc-item"><li><span><a href="#Inferences-by-features" data-toc-modified-id="Inferences-by-features-3.4.1"><span class="toc-item-num">3.4.1&nbsp;&nbsp;</span>Inferences by features</a></span><ul class="toc-item"><li><span><a href="#Temperature" data-toc-modified-id="Temperature-3.4.1.1"><span class="toc-item-num">3.4.1.1&nbsp;&nbsp;</span>Temperature</a></span></li><li><span><a href="#Fuel-Price" data-toc-modified-id="Fuel-Price-3.4.1.2"><span class="toc-item-num">3.4.1.2&nbsp;&nbsp;</span>Fuel Price</a></span></li><li><span><a href="#Markdown1-to-MArkdown-5" data-toc-modified-id="Markdown1-to-MArkdown-5-3.4.1.3"><span class="toc-item-num">3.4.1.3&nbsp;&nbsp;</span>Markdown1 to MArkdown 5</a></span></li><li><span><a href="#CPI" data-toc-modified-id="CPI-3.4.1.4"><span class="toc-item-num">3.4.1.4&nbsp;&nbsp;</span>CPI</a></span></li><li><span><a href="#unemployment" data-toc-modified-id="unemployment-3.4.1.5"><span class="toc-item-num">3.4.1.5&nbsp;&nbsp;</span>unemployment</a></span></li></ul></li></ul></li><li><span><a href="#Pair-Plots" data-toc-modified-id="Pair-Plots-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Pair Plots</a></span><ul class="toc-item"><li><span><a href="#Notable-Inferences" data-toc-modified-id="Notable-Inferences-3.5.1"><span class="toc-item-num">3.5.1&nbsp;&nbsp;</span>Notable Inferences</a></span></li></ul></li><li><span><a href="#Correlation-Heatmap" data-toc-modified-id="Correlation-Heatmap-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Correlation Heatmap</a></span></li><li><span><a href="#EDA-Final-words" data-toc-modified-id="EDA-Final-words-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>EDA Final words</a></span></li><li><span><a href="#Missing-Values" data-toc-modified-id="Missing-Values-3.8"><span class="toc-item-num">3.8&nbsp;&nbsp;</span>Missing Values</a></span><ul class="toc-item"><li><span><a href="#KNN-Imputation" data-toc-modified-id="KNN-Imputation-3.8.1"><span class="toc-item-num">3.8.1&nbsp;&nbsp;</span>KNN Imputation</a></span></li><li><span><a href="#Delete-Markdowns-from-dataset" data-toc-modified-id="Delete-Markdowns-from-dataset-3.8.2"><span class="toc-item-num">3.8.2&nbsp;&nbsp;</span>Delete Markdowns from dataset</a></span></li></ul></li><li><span><a href="#Other-data-cleaning-techniques" data-toc-modified-id="Other-data-cleaning-techniques-3.9"><span class="toc-item-num">3.9&nbsp;&nbsp;</span>Other data cleaning techniques</a></span></li></ul></li><li><span><a href="#Headsup-for-next-steps" data-toc-modified-id="Headsup-for-next-steps-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Headsup for next steps</a></span></li><li><span><a href="#Preparing-Test-dataset" data-toc-modified-id="Preparing-Test-dataset-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Preparing Test dataset</a></span></li><li><span><a href="#Model-1---No-data-transformation" data-toc-modified-id="Model-1---No-data-transformation-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Model 1 - No data transformation</a></span></li></ul></div>

## Problem Statement

This project is to predict the sales across different stores. 
The data containing historical sales data for 45 stores located in different regions - each store contains a
number of departments. The company also runs several promotional markdown events throughout the
year. These markdowns precede prominent holidays, the four largest of which are the Super Bowl, Labor
Day, Thanksgiving, and Christmas. The weeks including these holidays are weighted five times higher in
the evaluation than non-holiday weeks. Within the Excel Sheet, there are 3 Tabs – Stores, Features and
Sales

### Objective

The objective of this project is to 

1. Predict the department-wide sales for each store for the following year

2. Model the effects of markdowns on holiday weeks

3. Provide recommended actions based on the insights drawn, with prioritization placed on largest business
impact

## Data description

There are 3 datasets used in this project. Their respecitve details as follows:

### Stores

Anonymized information about the 45 stores, indicating the type and size of store

Store - Store ID

Type - Type of Store (A,B,C)

Size – Size of the store

### Features

Contains additional data related to the store, department, and regional activity for the given dates.

Store – Store ID

Date – Week start date

Temperature - average temperature in the region

Fuel_Price - cost of fuel in the region

MarkDown1-5 - anonymized data related to promotional markdowns. MarkDown data is only available
after Nov 2011, and is not available for all stores all the time. Any missing value is marked with NA

CPI - the consumer price index

Unemployment - the unemployment rate

IsHoliday - whether the week is a special holiday week

### Sales

Historical sales data, which covers to 2010-02-05 to 2012-11-01. Within this tab you will find the
following fields:

Store – Store ID

Dept – Department ID

Date – Week start date

Weekly_Sales -sales for the given department in the given store

IsHoliday - whether the week is a special holiday week

The dataset contains weekly sales data over the period of 3 years. Although data is clean for many features, there are mission values for Markdown columns anbd no data description about what these columns are. There could be challenges in understanding these columns and accommodating them to fit into the model along with handling lot of missing values there columns poccess

## Data Exploration

### Importing and Reading the datasets

The data is located in 3 csv files 

Sales - Retail Sales Prediction-Sales data-set.csv

Features - Retail Sales Prediction-Features data set.csv

Stores - Retail Sales Prediction-Stores data-set.csv

Overall steps for importing and reading the datasets are:

1. import pandas - very useful for the data analysis
2. Create a dictionary which holds the data read from all datasets - Dictionary is used to hold the data as a single object. This can be very convinient instead of creatig multiple variables
3. For each csv file, use pandas read.csv method by inputting the relevant file name

All the data read will be stored as 3 different dataframes within a single dictionary



In [1]:
import pandas as pd

In [4]:
# Store all the csv as python dataframes. Storing in single variable for convenience
data={
    "sales": pd.read_csv('Retail Sales Prediction-Sales data-set.csv'),
    "features": pd.read_csv('Retail Sales Prediction-Features data set.csv'),
    "stores": pd.read_csv('Retail Sales Prediction-Stores data-set.csv')
}

In [5]:

print(data['sales'].head())
print(len(data['sales']))

   Store  Dept        Date  Weekly_Sales  IsHoliday
0      1     1  05/02/2010      24924.50      False
1      1     1  12/02/2010      46039.49       True
2      1     1  19/02/2010      41595.55      False
3      1     1  26/02/2010      19403.54      False
4      1     1  05/03/2010      21827.90      False
421570


In [6]:

print(data['stores'].head())
print(len(data['stores']))

   Store Type    Size
0      1    A  151315
1      2    A  202307
2      3    B   37392
3      4    A  205863
4      5    B   34875
45


In [7]:

print(data['features'].head())
print(len(data['features']))

   Store        Date  Temperature  Fuel_Price  MarkDown1  MarkDown2  \
0      1  05/02/2010        42.31       2.572        NaN        NaN   
1      1  12/02/2010        38.51       2.548        NaN        NaN   
2      1  19/02/2010        39.93       2.514        NaN        NaN   
3      1  26/02/2010        46.63       2.561        NaN        NaN   
4      1  05/03/2010        46.50       2.625        NaN        NaN   

   MarkDown3  MarkDown4  MarkDown5         CPI  Unemployment  IsHoliday  
0        NaN        NaN        NaN  211.096358         8.106      False  
1        NaN        NaN        NaN  211.242170         8.106       True  
2        NaN        NaN        NaN  211.289143         8.106      False  
3        NaN        NaN        NaN  211.319643         8.106      False  
4        NaN        NaN        NaN  211.350143         8.106      False  
8190


#### Initial thoughts

Upon inspecting the data from 3 datasets, its evident that 

1. Sales data contains historical data
2. Stores contain the information regarding stores. This dataset doesn't add value on its own but combining store information with sales information helps in understanding the releation between sales across different type of stores
3. Features dataset contains sales information with aditional features. 

Upon carefully obeserving Features and Sales, its evident that there are few variables in Sales and Features in common and features has few addititional variables. Combining all 3 datasets helps in having a simplified single dataset and it helps leveraging more value out of the data.

##### Steps to merge the datasets

1. Join any 2 datasets based on the common columns. This can be similar to Left join in SQL
2. With the dataframe resulting from step 1, join the 3rd dataframe following similar approach as Step 1

In [8]:
# this df stores the merged df og features and stores csv
#data['merged'] = pd.merge(data['features'],data['stores'], on='Store',how='left')

In [9]:
# This df contains all 3 df joined int one
#data['merged_total'] = pd.merge(data['sales'], data['merged'],  how='left', left_on=['Store','Date','IsHoliday'], right_on = ['Store','Date','IsHoliday'])

In [10]:
#raw_data = data['merged_total']

##### Final dataset

As a result of merging all 3 dataframes, final dataset has a total of 421570 observations and 16 columns

In [11]:
#raw_data.describe()

### Save to file

Save the merged dataframe to a csv file and pickle for future reference.

#### CSV

File can be saved as csv using pandas to_csv method. 

In [12]:
#raw_data.to_csv('Retail_Sales_Merged.csv')

#### Pickle

Pickle is python's way to store the dataframes as python objects. Saving any dataframe to pickle helps in resuing the pickle object directly to read the dataframe to session

This file is stored on the disk and can be exported to any file storage sevice thus giving the flexibility resuing the object to read directly from storage locations when there is change in working environments.

Also, pickling helps to avoid all the data reading steps from csv and all steps involved in merging dataframes

In [13]:
raw_data = pd.read_pickle('Retail_Sales_Merged.pkl')
raw_data['Date'] = pd.to_datetime(raw_data['Date'])

### Data description

In [14]:
raw_data.describe().T

count           mean           std        min           25%  \
Store         421570.0      22.200546     12.785297      1.000     11.000000   
Dept          421570.0      44.260317     30.492054      1.000     18.000000   
Weekly_Sales  421570.0   15981.258123  22711.183519  -4988.940   2079.650000   
Temperature   421570.0      60.090059     18.447931     -2.060     46.680000   
Fuel_Price    421570.0       3.361027      0.458515      2.472      2.933000   
MarkDown1     150681.0    7246.420196   8291.221345      0.270   2240.270000   
MarkDown2     111248.0    3334.628621   9475.357325   -265.760     41.600000   
MarkDown3     137091.0    1439.421384   9623.078290    -29.100      5.080000   
MarkDown4     134967.0    3383.168256   6292.384031      0.220    504.220000   
MarkDown5     151432.0    4628.975079   5962.887455    135.160   1878.440000   
CPI           421570.0     171.201947     39.159276    126.064    132.022667   
Unemployment  421570.0       7.960289      1.863296      3.879      6.891000   
Size          421570.0  136727.915739  60980.583328  34875.000  93638.000000   

                       50%            75%            max  
Store             22.00000      33.000000      45.000000  
Dept              37.00000      74.000000      99.000000  
Weekly_Sales    7612.03000   20205.852500  693099.360000  
Temperature       62.09000      74.280000     100.140000  
Fuel_Price         3.45200       3.738000       4.468000  
MarkDown1       5347.45000    9210.900000   88646.760000  
MarkDown2        192.00000    1926.940000  104519.540000  
MarkDown3         24.60000     103.990000  141630.610000  
MarkDown4       1481.31000    3595.040000   67474.850000  
MarkDown5       3359.45000    5563.800000  108519.280000  
CPI              182.31878     212.416993     227.232807  
Unemployment       7.86600       8.572000      14.313000  
Size          140167.00000  202505.000000  219622.000000

### Univariate Analysis

In [15]:
import matplotlib
matplotlib.use('nbagg')
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

ModuleNotFoundError: No module named 'seaborn'

In [16]:
raw_data.dtypes

Store                    int64
Dept                     int64
Date            datetime64[ns]
Weekly_Sales           float64
IsHoliday                 bool
Temperature            float64
Fuel_Price             float64
MarkDown1              float64
MarkDown2              float64
MarkDown3              float64
MarkDown4              float64
MarkDown5              float64
CPI                    float64
Unemployment           float64
Type                    object
Size                     int64
dtype: object

Out of all numeric columns, Store, Dept and Size seem categorical variables. 

Out of all columns, below is the list of numeric columns

Temperature,
Fuel_Price,
MarkDown1,
MarkDown2,
MarkDown3,
MarkDown4,
MarkDown5,
CPI,
Unemployment

In [17]:
## USe interactive backend
matplotlib.interactive(True)

plt.ion()

In [18]:
for col in ['Temperature', 'Fuel_Price', 'MarkDown1', 'MarkDown2', 'MarkDown3', 'MarkDown4', 'MarkDown5', 'CPI', 'Unemployment']:
    fig, axs = plt.subplots(1, 1)
    axs.hist(raw_data[col].dropna())
    axs.set_title('Distribution plot for '+col)
    axs.set_xlabel(col)
    axs.set_ylabel('Frequency')

#### Inferences by features

##### Temperature
1. Data is normally distributed. The dataset contains sales data across seasons.
2. There are less number of outliers.
3. Few outliers are with 0-20 degrees
4. Depending on further analysis and while building model, decision to drop the outliers can be made

##### Fuel Price 
1. Distribution is not normal. At the same time, outliers are limited too
2. The data points are skewed.
3. Upon further analysis, if the impact of Fuel prices on Sales is high, there could be interesting insights drawn and more analysis on how the fuel prices are influencing the sales will be useful.

At this point in time, not much can be commented about this

##### Markdown1 to MArkdown 5
Data for all markdown fields is has common things in common
1. Heavily skewed data with outliers
2. It might be safe to assume to drop the outliers as their number is less.
3. Proper hyp[othesis has to be done before dropping them

##### CPI
There is a clear cluster in CPI distribution. Data can be assumed as categorical.

##### unemployment
1. Distribution is not normal. At the same time, outliers are limited too
2. The data points are skewed.
3. Upon further analysis, if the impact of Fuel prices on Sales is high, there could be interesting insights drawn and more analysis on how the unemployment is influencing the sales will be useful.



### Pair Plots

sns pairplots give the distribution as well as bivariate releationships between various columns

In [ ]:
sns.set()
cols = ['Weekly_Sales', 'Temperature', 'Fuel_Price', 'CPI', 'Unemployment']
sns.pairplot(raw_data[cols], size = 2.5)
plt.show();

#### Notable Inferences

1. Weekly sales look to be highly skewed. The sales are very high during some days and relatively constant for the rest of the days
2. CPI has 2 clusters
3. Temperature is almost normally distributed across data. Indicates the data is collected over all kinds of whether conditions

In [19]:
corrmat = raw_data.corr()
f, ax = plt.subplots(figsize=(12, 9))
sns.heatmap(corrmat, vmax=.8, square=True);

NameError: name 'sns' is not defined

### Correlation Heatmap

1. There is a strong correlation between MArkdown 1 and Markdown 4 . One of the columns might be safe tobe excluded when using regression models
2. Holiday and Markdown 2 and Markdown 3 have moderate correlation
3. Other columns looks to be fairly uncorrelated

In [20]:
from IPython.display import IFrame

Below is theinteractive visualization with 3 pages. More details can be found clicking on the next previous buttons or the header
Filters are also enabled and visualizations can be filtered by clicking on specific data points.

In [21]:

%%html
<div class='tableauPlaceholder' id='viz1569429520579' style='position: relative'><noscript><a href='#'><img alt=' ' src='https:&#47;&#47;public.tableau.com&#47;static&#47;images&#47;Re&#47;RetailSales_15694295079860&#47;RetailSalesvariableanalysis&#47;1_rss.png' style='border: none' /></a></noscript><object class='tableauViz'  style='display:none;'><param name='host_url' value='https%3A%2F%2Fpublic.tableau.com%2F' /> <param name='embed_code_version' value='3' /> <param name='site_root' value='' /><param name='name' value='RetailSales_15694295079860&#47;RetailSalesvariableanalysis' /><param name='tabs' value='no' /><param name='toolbar' value='yes' /><param name='static_image' value='https:&#47;&#47;public.tableau.com&#47;static&#47;images&#47;Re&#47;RetailSales_15694295079860&#47;RetailSalesvariableanalysis&#47;1.png' /> <param name='animate_transition' value='yes' /><param name='display_static_image' value='yes' /><param name='display_spinner' value='yes' /><param name='display_overlay' value='yes' /><param name='display_count' value='yes' /><param name='filter' value='publish=yes' /></object></div>                <script type='text/javascript'>                    var divElement = document.getElementById('viz1569429520579');                    var vizElement = divElement.getElementsByTagName('object')[0];                    vizElement.style.width='1016px';vizElement.style.height='991px';                    var scriptElement = document.createElement('script');                    scriptElement.src = 'https://public.tableau.com/javascripts/api/viz_v1.js';                    vizElement.parentNode.insertBefore(scriptElement, vizElement);                </script>

SyntaxError: invalid syntax (<ipython-input-21-ac6c86adb0df>, line 3)

### EDA Final words

Although not many patterns are found from varioud graphs above, one interesting trend observed is there are clear clusters of sales spikes during Nov-Dec for years 2010 and 2011. However, the pattern is missing for 2012 which is interesting. 

Markdowns are introduced only towards end of year 2011 (Nov-11-2011) and there seem to be no direct correlation between Markdown and Sales.
Markdowns also has many outliers.

One assumption to be made about Markdowns is to treat them as Promotions. But, there is no visible trend of Markdown vs Sales. 

Either, the assumption of Markdown being is wrong or the Promotions are not attractive or there could be other reasons which derived the lower sales in 2012

### Missing Values


In [22]:
raw_data.isnull().sum()

Store                0
Dept                 0
Date                 0
Weekly_Sales         0
IsHoliday            0
Temperature          0
Fuel_Price           0
MarkDown1       270889
MarkDown2       310322
MarkDown3       284479
MarkDown4       286603
MarkDown5       270138
CPI                  0
Unemployment         0
Type                 0
Size                 0
dtype: int64

Except markdowns, no other variable has missing values which is good.

For MArkdowns, almost half of the rows in the dataset has values missing.From the earlier analysis, the dataset has no markdowns until Nov-11-2011

Dropping the rows containing missing values may not be a good idea as it leads to information loss until Nov-11-2011

Other options like imputing the missing values using KNN may be considered. (Dropping the ides of other imputing techniques like mean or median substitution because of non existant data for complete period prior Nov-11-2011)

KNN imputing might be a reasonable technique in this scenario because it tries substitute the missing values based on nearest neighbours. If the data has a hidden pattern, then definitely substituting the missing values based on nearest neighbours would be more practical.

Other approcah is to drop the Markdowns completely and build model out od the remaining dataset. 

#### KNN Imputation

In [23]:
# Installing Fancy impute
#!pip install fancyimpute

#### Delete Markdowns from dataset

In [24]:
raw_data_no_markdown = raw_data.drop(["MarkDown1", "MarkDown2", "MarkDown3", "MarkDown4", "MarkDown5"], axis=1)

### Other data cleaning techniques

Leaving outliers as they are now. Decision on outliers can be taken when running the ML algorithms

Not performing any feature scalings as there are outliers in come of the columns. Feature scaling on dataset containing outliers may result in highly bisased datasets

Depending on the outcome of ML training, some of the variables may be converted to categorical and results may be compared.
Eg: Fuel price, CPI, Unemployment etc.,

Other data transformation technique is to make a column with weeknumbers

More data cleaning can be done based on the ML algorithm used as not all algorithms are sussiptible to outliers and feature sclales and their effects vary by algorithm

## Headsup for next steps

Continuing to the intuition from EDA, one option to make good quality predictions is to build mnultiple models

Some of them might be

1. Build different models for markdown impouted dataset and for datasets with no markdowns
2. As there is seasonality (Sales spikes at year end), it will be useful to build different models for Yearend alone (for all 3 years in single dataset) and different models for rest of the year (for all 3 years in single dataset)
3. Perform feature scaling and deal outliers where ever required (Eg: Random forests and decision trees doesn't require feature scaling. But linear regresion require feature scaling)
4. For models like linear regression, have to carefully look at correlation and perform PCA if needed.

Besides all these, it is worth to look at building the time series as there is some seasonality observed at year end for all 3 years

## Preparing Test dataset

From the datasets, weekly_sales is present for data prior to 26-11-2012. Post this date, there is no data for weekly_sales and hence data after this qualifies to be good candidate for the test dataset.

In [25]:
data['features']['Date'] = pd.to_datetime(data['features']['Date'])

In [26]:
no_weekly_sales_data = data['features'][data['features']['Date']>'2012-10-26']

In [27]:
test_data = pd.merge(no_weekly_sales_data,data['stores'], on='Store',how='left')

In [28]:
data['sales']['Date'] = pd.to_datetime(data['sales']['Date'])

In [29]:
test_data = pd.merge(test_data, data['sales'],  how='left', on=['Date'])

In [30]:
test_data.to_pickle('test_data_no_dept.pkl')

In [31]:
raw_data_no_dept = raw_data.drop('Dept',axis=1)

In [32]:
# Method to do basic data transforms and split the data
def transform_and_split(model_data, split_ratio=0.2):
    model_data['Store'].astype('category')
    model_data=pd.get_dummies(data=model_data,columns=['Type','IsHoliday'])
    model_data['Date']=model_data['Date'].dt.week
    sales = model_data['Weekly_Sales']
    features = model_data.drop('Weekly_Sales', axis = 1)
    # Shuffle and split the data into training and testing subsets
    X_train, X_test, y_train, y_test = train_test_split(features,sales,test_size=split_ratio, random_state=1)
    return X_train, X_test, y_train, y_test

## Model building

Approach to modelling:

1. Different models for Complete raw dataset (just a merge of all datasets provide) including markdown values and without any data transformation techniques other than replacing missing values of markdowns with zero.

2. Non transformed dataset as above but withour markdown columns

3. Apply some data transformations like replacing dates with week numbers and apply multiple algorithms


In [62]:
print(model1.train_error_metrics)
print(model1.error_metrics)

(0.09335136190344195, 4.4241154264012437e-13, 22072.269683556686)
(0.09415920290341961, 91.08942341878617, 21751.242044027294)


In [76]:
# Object to store all model details
all_models = pd.DataFrame(columns=['Model', 'Description', 'train_r2', 'validation_r2', 'train_mae', 'validation_mae', 'train_rmse', 'validation_rmse'])
model_array = []
# Utility function for saving model details
def store_model(modelInterface):
    model_data = {
        'Model': modelInterface.model,
        'Description': modelInterface.description,
        'train_r2': modelInterface.train_error_metrics[0],
        'validation_r2':modelInterface.error_metrics[0],
        'train_mae': modelInterface.train_error_metrics[1],
        'validation_mae': modelInterface.error_metrics[1],
        'train_rmse': modelInterface.train_error_metrics[2],
        'validation_rmse': modelInterface.error_metrics[2]
    }
    model_array.append(model_data)

In [34]:
import numpy as np
def rmse(predictions, targets):
    return np.sqrt(((predictions - targets) ** 2).mean())

In [35]:

def mae(predictions, targets):
    return np.abs(np.sum(predictions - targets))/len(predictions)

In [36]:
def compute_errors(model, y_test, X_test, do_print=False):
    error_r2 = r2_score(y_test, model.predict(X_test))
    error_mae = mae(y_test, model.predict(X_test))
    error_rmse = rmse(y_test, model.predict(X_test))
    if do_print:
        print('Model details: ')
        print(model)
        print('r2: ' + str(error_r2))
        print('mae: ' + str(error_mae))
        print('rmse: ' + str(error_rmse))
    return error_r2, error_mae, error_rmse

### Model 1.1 - No data transformation


In [37]:
raw_data_only_markdown = raw_data[raw_data['Date'] >  '2011-10-26']
raw_data_only_markdown.fillna(0,inplace=True)

/opt/conda/envs/fastai/lib/python3.6/site-packages/pandas/core/frame.py:3790: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  downcast=downcast, **kwargs)


In [38]:
raw_data_no_dept.fillna(0,inplace=True)

In [41]:
!pip install -U scikit-learn scipy matplotlib

    100% |████████████████████████████████| 6.7MB 6.8MB/s eta 0:00:01
    100% |████████████████████████████████| 25.2MB 2.3MB/s eta 0:00:01
    100% |████████████████████████████████| 13.1MB 4.7MB/s eta 0:00:01
    100% |████████████████████████████████| 296kB 40.7MB/s ta 0:00:01
  Found existing installation: scipy 1.2.0
    Uninstalling scipy-1.2.0:
      Successfully uninstalled scipy-1.2.0
  Found existing installation: matplotlib 3.0.2
    Uninstalling matplotlib-3.0.2:
      Successfully uninstalled matplotlib-3.0.2


In [42]:
import sklearn
from sklearn.model_selection import train_test_split


### Linear Regression

In [43]:
### Linear Regression
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score


In [44]:
class ModelInterface:
    def __init__(self, data, description ):
        self.data = data
        self.X_train, self.X_test, self.y_train, self.y_test = transform_and_split(data, 0.2)
        self.model = None
        self.description = description
        self.r2 = None
        self.mae = None
        self.rmse = None
        self.train_error_metrics: None
        self.error_metrics: None
    def transform_and_split(model_data, split_ratio=0.2):
        model_data['Store'].astype('category')
        model_data=pd.get_dummies(data=model_data,columns=['Type','IsHoliday'])
        model_data['Date']=model_data['Date'].dt.week
        sales = model_data['Weekly_Sales']
        features = model_data.drop('Weekly_Sales', axis = 1)
        # Shuffle and split the data into training and testing subsets
        X_train, X_test, y_train, y_test = train_test_split(features,sales,test_size=split_ratio, random_state=1)
        return X_train, X_test, y_train, y_test
    def rmse(predictions, targets):
        return np.sqrt(((predictions - targets) ** 2).mean())
    def mae(predictions, targets):
        return np.abs(np.sum(predictions - targets))/len(predictions)
    def compute_errors(self, model):
        error_r2 = r2_score(self.y_test, model.predict(self.X_test))
        error_mae = mae(self.y_test, model.predict(self.X_test))
        error_rmse = rmse(self.y_test, model.predict(self.X_test))
        return error_r2, error_mae, error_rmse
    def compute_train_errors(self, model):
        error_r2 = r2_score(self.y_train, model.predict(self.X_train))
        error_mae = mae(self.y_train, model.predict(self.X_train))
        error_rmse = rmse(self.y_train, model.predict(self.X_train))
        return error_r2, error_mae, error_rmse
    def runModel(self, model1):
        self.train_error_metrics = self.compute_train_errors(model1)
        self.error_metrics = self.compute_errors(model1)
    def printModel(self, model1):
        print('Model details: ')
        print(model1)
        print('Description: '+ self.description)
        print('train_r2: ' + str(self.train_error_metrics[0]))
        print('train_mae: ' + str(self.train_error_metrics[1]))
        print('train_rmse: ' + str(self.train_error_metrics[2]))
        print('validation_r2: ' + str(self.error_metrics[0]))
        print('validation_mae: ' + str(self.error_metrics[1]))
        print('validation_rmse: ' + str(self.error_metrics[2]))

In [45]:
model1 = ModelInterface(raw_data_only_markdown, "Linear regression Model with no data transformations. Includes all columns")
model1.model = LinearRegression(normalize=True).fit(model1.X_train,model1.y_train)
model1.runModel(model1.model)
model1.printModel(model1.model)

Model details: 
LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=True)
Description: Linear regression Model with no data transformations. Includes all columns
train_r2: 0.09335136190344195
train_mae: 4.4241154264012437e-13
train_rmse: 22072.269683556686
validation_r2: 0.09415920290341961
validation_mae: 91.08942341878617
validation_rmse: 21751.242044027294


### Grid Search

In [46]:
from sklearn.metrics import make_scorer
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import ShuffleSplit

def fit_model(algorithm, X, y, params):
    """ Performs grid search over the 'max_depth' parameter for a 
        decision tree regressor trained on the input data [X, y]. """
    
    # Create cross-validation sets from the training data
    cv_sets = ShuffleSplit(n_splits = 10, test_size = 0.20, random_state = 0)

    # Create a decision tree regressor object
    regressor = algorithm

    # Transform 'performance_metric' into a scoring function using 'make_scorer' 
    scoring_fnc = make_scorer(r2_score)

    # Create the grid search object
    grid = GridSearchCV(regressor, param_grid= params, scoring=scoring_fnc, cv=cv_sets)

    # Fit the grid search object to the data to compute the optimal model
    grid = grid.fit(X, y)

    # Return the optimal model after fitting the data
    return grid.best_estimator_

### Decision Trees

In [47]:
from sklearn.tree import DecisionTreeRegressor

In [48]:
model2 = ModelInterface(raw_data_only_markdown, "Decision Tree regression Model with no data transformations. Includes all columns")
model2.model = DecisionTreeRegressor(criterion='mse', max_depth=7, max_features=None,
           max_leaf_nodes=None, min_impurity_decrease=0.0,
           min_impurity_split=None, min_samples_leaf=1,
           min_samples_split=2, min_weight_fraction_leaf=0.0,
           presort=False, random_state=None, splitter='best').fit(model2.X_train,model2.y_train)
model2.runModel(model2.model)
model2.printModel(model2.model)

Model details: 
DecisionTreeRegressor(criterion='mse', max_depth=7, max_features=None,
                      max_leaf_nodes=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      presort=False, random_state=None, splitter='best')
Description: Decision Tree regression Model with no data transformations. Includes all columns
train_r2: 0.7608430604905764
train_mae: 3.4509025873090033e-12
train_rmse: 11336.238663526812
validation_r2: 0.7591936136791231
validation_mae: 51.81069297940961
validation_rmse: 11214.815241125414


In [49]:
from sklearn.ensemble import RandomForestRegressor
model3 = ModelInterface(raw_data_only_markdown, "Random forest regression Model with no data transformations. Includes all columns")
model3.model = RandomForestRegressor(n_estimators=100, criterion='mse', min_samples_split=2, min_samples_leaf=1, min_weight_fraction_leaf=0.0, max_features='auto', max_leaf_nodes=None, min_impurity_decrease=0.0, min_impurity_split=None, bootstrap=True, oob_score=True, n_jobs=None, random_state=1, verbose=0, warm_start=False).fit(model3.X_train, model3.y_train)
model3.runModel(model3.model)
model3.printModel(model3.model)

Model details: 
RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
                      max_features='auto', max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, n_estimators=100,
                      n_jobs=None, oob_score=True, random_state=1, verbose=0,
                      warm_start=False)
Description: Random forest regression Model with no data transformations. Includes all columns
train_r2: 0.9949871119356408
train_mae: 19.208004602480617
train_rmse: 1641.2377235868435
validation_r2: 0.969933597866702
validation_mae: 52.08788675189235
validation_rmse: 3962.7698842137193


In [50]:
model4 = ModelInterface(raw_data_no_markdown, "Linear regression Model with no data transformations. Includes all columns except markdowns")
model4.model = LinearRegression(normalize=True).fit(model4.X_train,model4.y_train)
model4.runModel(model4.model)
model4.printModel(model4.model)

Model details: 
LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=True)
Description: Linear regression Model with no data transformations. Includes all columns except markdowns
train_r2: 0.08833306322917589
train_mae: 18.379639828498238
train_rmse: 21675.247811126523
validation_r2: 0.092406011546733
validation_mae: 76.49072870460421
validation_rmse: 21674.82239013148


In [51]:
model5 = ModelInterface(raw_data_no_markdown, "Decision Tree regression Model with no data transformations. Includes all columns except markdowns")
model5.model = DecisionTreeRegressor(criterion='mse', max_depth=7, max_features=None,
           max_leaf_nodes=None, min_impurity_decrease=0.0,
           min_impurity_split=None, min_samples_leaf=1,
           min_samples_split=2, min_weight_fraction_leaf=0.0,
           presort=False, random_state=None, splitter='best').fit(model5.X_train,model5.y_train)
model5.runModel(model5.model)
model5.printModel(model5.model)

Model details: 
DecisionTreeRegressor(criterion='mse', max_depth=7, max_features=None,
                      max_leaf_nodes=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      presort=False, random_state=None, splitter='best')
Description: Decision Tree regression Model with no data transformations. Includes all columns except markdowns
train_r2: 0.7629119032567171
train_mae: 1.827161631005865e-11
train_rmse: 11053.526181981952
validation_r2: 0.7719734240466052
validation_mae: 11.046723206674706
validation_rmse: 10864.318152665051


In [52]:
model6 = ModelInterface(raw_data_no_markdown, "Random forest regression Model with no data transformations. Includes all columns except markdowns")
model6.model = RandomForestRegressor(n_estimators=100, criterion='mse', 
                                     min_samples_split=2, min_samples_leaf=1, 
                                     min_weight_fraction_leaf=0.0, max_features='auto', 
                                     max_leaf_nodes=None, min_impurity_decrease=0.0, 
                                     min_impurity_split=None, bootstrap=True, oob_score=True, 
                                     n_jobs=None, random_state=1, verbose=0, warm_start=False).fit(model6.X_train, model6.y_train)
model6.runModel(model6.model)
model6.printModel(model6.model)

Model details: 
RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
                      max_features='auto', max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, n_estimators=100,
                      n_jobs=None, oob_score=True, random_state=1, verbose=0,
                      warm_start=False)
Description: Random forest regression Model with no data transformations. Includes all columns except markdowns
train_r2: 0.9954672133720925
train_mae: 2.3539400974334352
train_rmse: 1528.3703612895772
validation_r2: 0.9688096528152962
validation_mae: 21.575057240790354
validation_rmse: 4018.091725050686


In [53]:
model7 = ModelInterface(raw_data_no_dept, "Decision Tree regression Model with no data transformations. Includes all columns including markdowns. This dataset contains data prior to the introduction of markdowns and also later")
model7.model = DecisionTreeRegressor(criterion='mse', max_depth=7, max_features=None,
           max_leaf_nodes=None, min_impurity_decrease=0.0,
           min_impurity_split=None, min_samples_leaf=1,
           min_samples_split=2, min_weight_fraction_leaf=0.0,
           presort=False, random_state=None, splitter='best').fit(model7.X_train,model7.y_train)
model7.runModel(model7.model)
model7.printModel(model7.model)

Model details: 
DecisionTreeRegressor(criterion='mse', max_depth=7, max_features=None,
                      max_leaf_nodes=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      presort=False, random_state=None, splitter='best')
Description: Decision Tree regression Model with no data transformations. Includes all columns including markdowns. This dataset contains data prior to the introduction of markdowns
train_r2: 0.09656254543257947
train_mae: 9.401427121469734e-12
train_rmse: 21577.19641228585
validation_r2: 0.09711426429031167
validation_mae: 73.92400507043483
validation_rmse: 21618.528916645013


In [59]:
model8 = ModelInterface(raw_data_no_dept, "Random forest regression Model with no data transformations. This dataset contains data prior to the introduction of markdowns and also later")
model8.model = RandomForestRegressor(n_estimators=100, criterion='mse', 
                                     min_samples_split=2, min_samples_leaf=1, 
                                     min_weight_fraction_leaf=0.0, max_features='auto', 
                                     max_leaf_nodes=None, min_impurity_decrease=0.0, 
                                     min_impurity_split=None, bootstrap=True, oob_score=True, 
                                     n_jobs=None, random_state=1, verbose=0, warm_start=False).fit(model8.X_train, model8.y_train)
model8.runModel(model8.model)
model8.printModel(model8.model)

Model details: 
RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
                      max_features='auto', max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, n_estimators=100,
                      n_jobs=None, oob_score=True, random_state=1, verbose=0,
                      warm_start=False)
Description: Random forest regression Model with no data transformations. This dataset contains data prior to the introduction of markdowns and also later
train_r2: 0.10625597624689775
train_mae: 1.2867351386963013
train_rmse: 21461.127985366205
validation_r2: 0.07716769995044315
validation_mae: 83.57151518176187
validation_rmse: 21856.022824493226


In [98]:
raw_data_no_markdown_week_num = raw_data_no_markdown
raw_data_no_markdown['WeekNumber'] = raw_data_no_markdown['Date'].dt.week

In [99]:
model9 = ModelInterface(raw_data_no_markdown_week_num, "Random forest regression Model with no data transformations. This dataset contains data prior to the introduction of markdowns and also later")
model9.model = RandomForestRegressor(n_estimators=100, criterion='mse', 
                                     min_samples_split=2, min_samples_leaf=1, 
                                     min_weight_fraction_leaf=0.0, max_features='auto', 
                                     max_leaf_nodes=None, min_impurity_decrease=0.0, 
                                     min_impurity_split=None, bootstrap=True, oob_score=True, 
                                     n_jobs=None, random_state=1, verbose=0, warm_start=False).fit(model9.X_train, model9.y_train)
model9.runModel(model9.model)
model9.printModel(model9.model)

Model details: 
RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
                      max_features='auto', max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, n_estimators=100,
                      n_jobs=None, oob_score=True, random_state=1, verbose=0,
                      warm_start=False)
Description: Random forest regression Model with no data transformations. This dataset contains data prior to the introduction of markdowns and also later
train_r2: 0.9955032539157406
train_mae: 1.9686046922812663
train_rmse: 1522.2821385661111
validation_r2: 0.96894874494347
validation_mae: 21.493581191735636
validation_rmse: 4009.1224532362717


In [58]:
all_models_pickle = {
    'model1': model1,
    'model2': model2,
    'model3': model3,
    'model4': model4,
    'model5': model5,
    'model6': model6,
    'model7': model7,
    'model8': model8
}

In [56]:
import pickle

with open('all_models.pickle', 'wb') as handle:
    pickle.dump(all_models_pickle, handle, protocol=pickle.HIGHEST_PROTOCOL)

In [ ]:
#with open('all_models.pickle', 'rb') as handle:
    #all_models_pickle = pickle.load(handle)

In [80]:
store_model(model1)
store_model(model2)
store_model(model3)
store_model(model4)
store_model(model5)
store_model(model6)
store_model(model7)
store_model(model8)
store_model(model9)

In [105]:
all_models=pd.DataFrame(model_array)

In [109]:
all_models.to_csv('all_models.csv')

In [90]:
all_models = all_models[['Description', 'Model', 'train_r2', 'validation_r2', 'train_mae', 
       'validation_mae', 'train_rmse', 'validation_rmse']]

In [97]:
raw_data_no_markdown_week_num = raw_data_no_markdown
raw_data_no_markdown['WeekNumber'] = raw_data_no_markdown['Date'].dt.week

## Modelling Stage Final thoughts